cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll

```diff
@@ -327,6 +327,11 @@ class CallAllocationExpr extends AllocationExpr, FunctionCall {
 
   override Expr getReallocPtr() { result = getArgument(target.getReallocPtrArg()) }
 
+  override Type getAllocatedElementType() {
+    result = this.getFullyConverted().getType().(PointerType).getBaseType() and
```

Comment:
I think this needs to be `getUnspecifiedType()` rather than `getType()` to expand typedefs and strip off `const` or `volatile` specifiers before the inline cast to `PointerType`.

Reply:
I reassessed this, and `getType()` is wrong, as it doesn't handle typedefs correctly.

However, I think the correct predicate is `getUnderlyingType()`, as I don't see why we would want to strip of `const` or `volatile` members.

If we would want to do that, we'd need to also adjust the other instances, as `NewArrayExpr.getAllocatedElementType()` does not strip any modifiers either.

Reply:
I think `getType().stripTopLevelSpecifiers()` is the right one to use here. It resolves typedefs and removes specifiers only at the outermost layer. Everything below the `PointerType` will be left alone.

Reply:
@rdmarsh2 what do you think?

Reply:
Yeah, `stripTopLevelSpecifiers()` is the best for this situation. We need to strip off specifiers because we represent them with a separate class, `SpecifiedType`, which wraps other instances of `Type`. For instance, a `const int * const` will be represented by a chain of four `Type` instances: a `SpecifiedType` wrapping a ` PointerType` wrapping a second `SpecifiedType`, which wraps an `IntType`. Only the second instance can be cast to `PointerType`. `stripTopLevelSpecifiers()` does exactly enough work to reach the `PointerType`, without removing the second `SpecifiedType`.